src/Ouroboros-Mini_Protocols-Chain_Sync.thy

```diff
@@ -124,30 +135,43 @@ definition index :: "('i \<Rightarrow> 'q) \<Rightarrow> 'q \<Rightarrow> 'i lis
 definition first_intersection_point :: "('i \<Rightarrow> 'q) \<Rightarrow> 'q list \<Rightarrow> 'i list \<rightharpoonup> 'q" where
   [simp]: "first_intersection_point \<psi> qs C  = find (\<lambda>q. q \<in> \<psi> ` set C) qs"
 
+definition last_intersection_point :: "('i \<Rightarrow> 'q) \<Rightarrow> 'i list \<Rightarrow> 'i list \<Rightarrow> 'q" where
+  [simp]: "last_intersection_point \<psi> C C' = (ARG_MAX (\<lambda>q. index \<psi> q C) q. q \<in> \<psi> ` (set C \<inter> set C'))"
+
 corec server_program where
-  "server_program \<psi> C k b =
+  "server_program \<psi> C k b a =
     \<down> M. (partial_case M of
       Done \<Rightarrow>
         \<bottom> |
       Cont (FindIntersect qs) \<Rightarrow>
         (case first_intersection_point \<psi> qs C of
           None \<Rightarrow>
             \<up> Cont IntersectNotFound;
-            server_program \<psi> C k b |
+            server_program \<psi> C k b a |
           Some q \<Rightarrow>
             \<up> Cont (IntersectFound q);
-            server_program \<psi> C (index \<psi> q C) True
+            server_program \<psi> C (index \<psi> q C) True a
         ) |
       Cont RequestNext \<Rightarrow>
         if b then
           \<up> Cont (RollBackward (\<psi> (C ! k)));
-          server_program \<psi> C k False
+          server_program \<psi> C k False a
         else if Suc k < length C then
           \<up> Cont (RollForward (C ! Suc k));
-          server_program \<psi> C (Suc k) b
-        else \<comment> \<open>only for this initial implementation\<close>
+          server_program \<psi> C (Suc k) b a
+        else
           \<up> Cont AwaitReply;
-          server_program \<psi> C k b
+          a \<rightarrow> C'. (
+            \<comment> \<open>assumed that at least \<^term>\<open>C\<close>~and~\<^term>\<open>C'\<close> coincide in the genesis block\<close>
+            let q = last_intersection_point \<psi> C C' in
+            if q = \<psi> (last C) then
+              \<comment> \<open>assumed that \<^term>\<open>C'\<close> is longer than \<^term>\<open>C\<close>\<close>
```

Comment:
If we don't assume that then there is the possibility of `C'` being exactly the same as `C`. In this case, what message should we send to the client? Perhaps a message instructing the client to roll back to the current point (and thus resulting in no change at all) might do the trick. :man_shrugging:

Reply:
Well, if we implement the chain update on the server side via shared-memory emulation, it would be common for the same chain to be sent. This would usually indicated that the chain wasn’t changed at all. What would we do in this case? Well, just wait further.

Reply:
Fixed as part of 4b6488c47e1500f3f69dfab23af37680da18676a, please check.